client/lib/url/is-redirect-allowed.ts

```diff
@@ -0,0 +1,64 @@
+/**
+ * Return false for absolute URLs which are on unknown hosts.
+ *
+ * Because the `redirectTo` query param on the pending page is the target of
+ * that page's redirect, we want to make sure we do not create an open redirect
+ * security hole that could go anywhere. This function will disallow a URL
+ * which is absolute and on an unknown host.
+ */
+
+// TODO: Replace original implementation of isRedirectAllowed with a reference to this shared module
+export function isRedirectAllowed( url: string, siteSlug: string | undefined ): boolean {
+	if ( url.startsWith( '/' ) ) {
+		return true;
+	}
+
+	const allowedHostsForRedirect = [
+		'wordpress.com',
+		'wpcalypso.wordpress.com',
+		'horizon.wordpress.com',
+		'calypso.localhost',
+		'jetpack.cloud.localhost',
+		'cloud.jetpack.com',
+		'jetpack.com',
+		'akismet.com',
+		siteSlug,
+	];
+
+	try {
+		const parsedUrl = new URL( url );
+		const { hostname, pathname } = parsedUrl;
+		if ( ! hostname ) {
+			return false;
+		}
+
+		// For subdirectory site, check that both hostname and subdirectory matches
+		// the siteSlug (host.name::subdirectory).
+		if ( siteSlug?.includes( '::' ) ) {
+			const [ hostnameFromSlug, ...subdirectoryParts ] = siteSlug.split( '::' );
+			const subdirectoryPathFromSlug = subdirectoryParts.join( '/' );
+			if (
+				hostname !== hostnameFromSlug &&
+				! pathname?.startsWith( `/${ subdirectoryPathFromSlug }` )
+			) {
+				return false;
+			}
+			return true;
+		}
```

Comment:
Do you have specific examples in mind where this `siteSlug` logic would be used?  I wonder if it would be better to simplify the implementation of `isRedirectAllowed ` for now by removing the param `siteSlug` (and this logic), and only implement when it's needed.

Reply:
Great question! I don't have a specific example, but I was planning on replacing the original implementation of [isRedirectAllowed](https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/checkout/src/lib/pending-page.ts#L244) soon which is why I kept it intact.

I was hoping to separate that refactor into its own PR for testing and easier reversion ( because it touches checkout code )

What I'm happy to do is remove it for now and add it back in when I create the follow-up PR 🙂

Reply:
[e2f5c73](https://github.com/Automattic/wp-calypso/pull/92106/commits/e2f5c73a797cc2a65b8db4bd5717b60282f0d086)